core/store/orm/orm.go

```diff
@@ -291,41 +295,123 @@ func (orm *ORM) convenientTransaction(callback func(*gorm.DB) error) error {
 	return postgres.GormTransaction(context.Background(), orm.DB, callback)
 }
 
-// SaveJobRun updates UpdatedAt for a JobRun and saves it
+// SaveJobRun updates UpdatedAt for a JobRun and updates its status, finished
+// at and run results.
+// It auto-inserts run results if none are existing already for the given
+// jobrun/taskruns.
 func (orm *ORM) SaveJobRun(run *models.JobRun) error {
+	if run.ID == uuid.Nil {
+		return errors.New("job run did not have an ID set")
+	}
 	if err := orm.MustEnsureAdvisoryLock(); err != nil {
 		return err
 	}
-	// FIXME: Be neater about this context timeout
-	ctx, cancel := context.WithTimeout(context.Background(), 15*time.Second)
+	ctx, cancel := postgres.DefaultQueryCtx()
 	defer cancel()
-	return postgres.GormTransaction(ctx, orm.DB, func(dbtx *gorm.DB) error {
-		result := dbtx.Unscoped().
-			//Session(&gorm.Session{FullSaveAssociations: true}). // We want to save the RunResult and TaskRuns also.
-			Where("updated_at = ?", run.UpdatedAt).
-			Omit("deleted_at").
-			Save(run)
+	err := postgres.GormTransaction(ctx, orm.DB, func(dbtx *gorm.DB) error {
+		result := dbtx.Exec(`
+UPDATE job_runs SET "status"=?, "finished_at"=?, "updated_at"=NOW(), "creation_height"=?, "observed_height"=?, "payment"=?
+WHERE updated_at = ? AND "id" = ?`,
+			run.Status, run.FinishedAt, run.CreationHeight, run.ObservedHeight, run.Payment,
+			run.UpdatedAt, run.ID,
+		)
 		if result.Error != nil {
-			if strings.Contains(result.Error.Error(), "duplicate key value violates unique constraint") {
-				return ErrOptimisticUpdateConflict
-			}
-			return result.Error
+			return errors.Wrap(result.Error, "failed to update job run")
 		}
 		if result.RowsAffected == 0 {
 			return ErrOptimisticUpdateConflict
 		}
-		err := dbtx.Unscoped().Save(&run.Result).Error
-		if err != nil {
-			return err
+
+		uRunResultStrs := []string{}
+		uRunResultArgs := []interface{}{}
+		if run.Result.ID > 0 {
```

Comment:
is run.Result guaranteed to be preloaded here?

Reply:
I don't know, but if it isn't then we will return an error on line 332.